[JSC] Fix race condition in Atomics.{wait,waitAsync}

jjgriego opened the pull request:
#### bb3be5c8f260f4dbdece34885f9de06dd5254b6d
<pre>
[JSC] Fix race condition in Atomics.{wait,waitAsync}
<a href="https://bugs.webkit.org/show_bug.cgi?id=248631">https://bugs.webkit.org/show_bug.cgi?id=248631</a>

Reviewed by Yusuke Suzuki.

At the moment, this is manifesting as intermittent timeouts in the arm32 linux
test runner, for example: <a href="https://build.webkit.org/#/builders/24/builds/17430">https://build.webkit.org/#/builders/24/builds/17430</a>

After investigation, this is the test `stress/waitasync-waiter-list-order.js`
hanging; this manifests as some of the workers hanging on the initial
`Atomic.wait` call in that test--using a debugger to examine the memory being
waited on reveals that they should all have woken up and moved on.

The implementation of both `Atomics.wait` and `Atomics.waitAsync` begin by
atomically reading from the array and comparing the result to the expected
value, then obtaining a lock on the list of waiters and proceeding from
there--this is contrary to the spec [1], which requires that the atomic read
happens while the list of waiters is exclusively locked and allows a race where
the following interleaving results in A hanging indefinitely if no other calls
to Atomic.notify on that address occur:

    thread A: begin Atomic.wait(buffer, idx, expected_value)
    thread A: read buffer[idx], get expected_value

    thread B: perform Atomic.store(buffer, idx, unexpected_value)
    thread B: perform Atomic.notify(buffer, idx) [nothing is notified since the list of waiters is empty]

    thread A: lock list of waiters, wait on condition variable, hang waiting for next notify

----

This is fixed by delaying the atomic read until the lock on the list of waiters
is held

[1] <a href="https://tc39.es/ecma262/multipage/structured-data.html#sec-atomics.wait">https://tc39.es/ecma262/multipage/structured-data.html#sec-atomics.wait</a>

* Source/JavaScriptCore/runtime/AtomicsObject.cpp:
(JSC::atomicsWaitImpl):
* Source/JavaScriptCore/runtime/WaiterListManager.cpp:
(JSC::WaiterListManager::waitImpl):
(JSC::WaiterListManager::wait):
(JSC::WaiterListManager::timeoutAsyncWaiter):
(JSC::WaiterListManager::addAsyncWaiter): Deleted.
* Source/JavaScriptCore/runtime/WaiterListManager.h:

Canonical link: <a href="https://commits.webkit.org/257423@main">https://commits.webkit.org/257423@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/698248f6519bb13180c5081c517d9ce9177d82ea

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/98965 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/8172 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/32126 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/108376 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/168629 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/8726 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/85526 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/91494 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/106344 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/104671 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/6607 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/90193 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/33637 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/88429 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/21538 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/76491 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/89698 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/2076 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/23054 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/85509 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/1982 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/85526 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/28815 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/6942 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/42528 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/88364 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2584 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/3391 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/19777 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.h:
```diff
@@ -202,13 +202,41 @@ class WaiterList : public ThreadSafeRefCounted<WaiterList> {
+private:
+    WaiterListManager& m_manager;
+    Ref<WaiterList> m_list;
+    Locker<WTF::Lock> m_locker;
```
Locker should not be in a member of some objects since this is really dangerous idiom for locking.

Constellation commented on Source/JavaScriptCore/runtime/AtomicsObject.cpp:
```diff
@@ -455,8 +456,52 @@ JSValue atomicsWaitImpl(JSGlobalObject* globalObject, JSArrayType* typedArray, u
+
+    default:
+        RELEASE_ASSERT_NOT_REACHED();
+    }
```
I think we should move these code in `WaiterListManager`. Extracting these part from `WaiterListManager` is making lock management harder (e.g., `LockedWaiterList` should not have Locker field.)

Constellation commented on Source/JavaScriptCore/runtime/AtomicsObject.cpp:
```diff
@@ -455,8 +456,52 @@ JSValue atomicsWaitImpl(JSGlobalObject* globalObject, JSArrayType* typedArray, u
 
-    AtomicsWaitValidation validation = WTF::atomicLoad(ptr) == expectedValue ? AtomicsWaitValidation::Pass : AtomicsWaitValidation::Fail;
-    return WaiterListManager::singleton().wait(globalObject, vm, ptr, validation, timeout, type);
+    switch (type) {
```
I think we can just have `wait32` and `wait64`, based on type of `ptr`.

Constellation replied:
Or, we can pass width, 4 or 8. and read based on that passed size.

Constellation reviewed (changes requested):
Commented.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.cpp:
```diff
@@ -50,38 +50,54 @@ WaiterListManager& WaiterListManager::singleton()
+            else {
+                isAsync = true;
+                JSPromise* promise = JSPromise::create(vm, globalObject->promiseStructure());
+                Ref<Waiter> waiter = adoptRef(*new Waiter(promise));
```
Do not perform GC allocation after taking a lock. It is unfortunate, but we should allocate promise before taking a lock.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.cpp:
```diff
@@ -50,38 +50,54 @@ WaiterListManager& WaiterListManager::singleton()
-        return vm.smallStrings.notEqualString();
+            object->putDirect(vm, vm.propertyNames->async, jsBoolean(isAsync));
+            object->putDirect(vm, vm.propertyNames->value, value);
+            return object;
```
Do these things after releasing lock, since it can have potential GC allocations.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.cpp:
```diff
@@ -101,49 +117,44 @@ JSValue WaiterListManager::wait(JSGlobalObject* globalObject, VM& vm, void* ptr,
-                    notifyWaiterImpl(NoLockingNecessary, WTFMove(waiter), ResolveResult::Timeout);
-            });
-            waiter->setTimer(listLocker, WTFMove(timer));
+void WaiterListManager::timeoutAsyncWaiter(void* ptr, Ref<Waiter> waiter)
```
Take `Ref<Waiter>&&`.

Constellation reviewed (approved):
r=me with comments.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.cpp:
```diff
@@ -50,38 +50,56 @@ WaiterListManager& WaiterListManager::singleton()
+
+        Ref<WaiterList> list = findOrCreateList(ptr);
+        JSPromise* promise = JSPromise::create(vm, globalObject->promiseStructure());
+        Ref<Waiter> waiter = adoptRef(*new Waiter(promise));
```
You should not create waiter when it is not added to the list.

Constellation reviewed (changes requested):
Commented about the bug

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.cpp:
```diff
@@ -50,38 +50,56 @@ WaiterListManager& WaiterListManager::singleton()
+            }
         }
 
+        dataLogLnIf(WaiterListsManagerInternal::verbose && isAsync, "WaiterListManager added a new AsyncWaiter ", RawPointer(waiter.ptr()), " to a waiterList for ptr ", RawPointer(ptr));
```
This looks not so great, in particular, `isAsync`.
You should only use `WaiterListsManagerInternal::verbose` for disable / enable of this dataLog.

jjgriego committed:
[JSC] Fix race condition in Atomics.{wait,waitAsync}
https://bugs.webkit.org/show_bug.cgi?id=248631

Reviewed by Yusuke Suzuki.

At the moment, this is manifesting as intermittent timeouts in the arm32 linux
test runner, for example: https://build.webkit.org/#/builders/24/builds/17430

After investigation, this is the test `stress/waitasync-waiter-list-order.js`
hanging; this manifests as some of the workers hanging on the initial
`Atomic.wait` call in that test--using a debugger to examine the memory being
waited on reveals that they should all have woken up and moved on.

The implementation of both `Atomics.wait` and `Atomics.waitAsync` begin by
atomically reading from the array and comparing the result to the expected
value, then obtaining a lock on the list of waiters and proceeding from
there--this is contrary to the spec [1], which requires that the atomic read
happens while the list of waiters is exclusively locked and allows a race where
the following interleaving results in A hanging indefinitely if no other calls
to Atomic.notify on that address occur:

    thread A: begin Atomic.wait(buffer, idx, expected_value)
    thread A: read buffer[idx], get expected_value

    thread B: perform Atomic.store(buffer, idx, unexpected_value)
    thread B: perform Atomic.notify(buffer, idx) [nothing is notified since the list of waiters is empty]

    thread A: lock list of waiters, wait on condition variable, hang waiting for next notify

----

This is fixed by delaying the atomic read until the lock on the list of waiters
is held

[1] https://tc39.es/ecma262/multipage/structured-data.html#sec-atomics.wait

* Source/JavaScriptCore/runtime/AtomicsObject.cpp:
(JSC::atomicsWaitImpl):
* Source/JavaScriptCore/runtime/WaiterListManager.cpp:
(JSC::WaiterListManager::waitImpl):
(JSC::WaiterListManager::wait):
(JSC::WaiterListManager::timeoutAsyncWaiter):
(JSC::WaiterListManager::addAsyncWaiter): Deleted.
* Source/JavaScriptCore/runtime/WaiterListManager.h:

Canonical link: https://commits.webkit.org/257423@main
